src/nats.c

```diff
@@ -760,41 +763,62 @@ _asyncCbsThread(void *arg)
         sc = cb->sc;
 #endif
 
+        // callback handlers can be updated on a live connection, so we need to
+        // lock.
+        natsMutex_Lock(nc->mu);
```

Comment:
You would also need to reset cbHandler/Closure and errHandler/closure in each loop iteration!